File: mobile/lib/models/Reservation.dart
import 'package:mobile/models/Device.dart';
import 'package:mobile/models/User.dart';

class Reservation {
  final String id;

  final User user;
  final Device device;

  final DateTime startedAt;
  final DateTime endedAt;

  final DateTime? createdAt;
  final DateTime? updatedAt;

  Reservation(
      {required this.id,
      required this.user,
      required this.device,
      required this.startedAt,
      required this.endedAt,
      this.createdAt,
      this.updatedAt});

  factory Reservation.fromJson(Map<String, dynamic> json) {
    return Reservation(
        id: json["_id"],
        user: User.fromJson(json["user"]),
        device: Device.fromJson(json["device"]),
        startedAt: DateTime.parse(json["startedAt"]),
        endedAt: DateTime.parse(json["endedAt"]),
        createdAt: json["createdAt"] != null
            ? DateTime.parse(json["createdAt"])
            : null,
        updatedAt: json["updatedAt"] != null
            ? DateTime.parse(json["updatedAt"])
            : null);
  }
}

File: backend/src/main/java/ai/threeeye/yeyak/domain/Reservation.java
package ai.threeeye.yeyak.domain;

import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.UUID;

@Data
@Builder
@ToString
@NoArgsConstructor
@AllArgsConstructor
@Document("reservations")
public class Reservation {

    @Id
    @Builder.Default
    private String _id = UUID.randomUUID().toString();

    @DBRef
    private User user;
    @DBRef
    private Device device;

    private LocalDateTime startedAt;
    private LocalDateTime endedAt;

    @CreatedDate
    private LocalDateTime createdAt;
    @LastModifiedDate
    private LocalDateTime updatedAt;

}
